src-ui.v3/src/Main.hs

```diff
@@ -731,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
 joinE :: forall t m a. (Reflex t, MonadHold t m) => Event t (Event t a) -> m (Event t a)
 joinE = fmap switch . hold never
 
-button_ :: forall t m. (DomBuilder t m, PostBuild t m) => T.Text -> m (Event t ())
-button_ t = do
+clickElement_ :: forall t m. (DomBuilder t m, PostBuild t m) => T.Text -> T.Text -> m (Event t ())
+clickElement_ elm t = do
   let cfg = (def :: ElementConfig EventResult t (DomBuilderSpace m))
         & elementConfig_eventSpec %~ addEventSpecFlags (Proxy :: Proxy (DomBuilderSpace m)) Click (\_ -> preventDefault)
-  (e, _) <- element "button" cfg $ text t
+  (e, _) <- element elm cfg $ text t
   pure $ domEvent Click e
 
+stripSearch :: JSS.JSString -> JSS.JSString
+stripSearch sJ
+  | Just sJ'  <- JSS.stripPrefix "^" sJ = sJ'
+  | Just sJ' <- JSS.stripSuffix "$" sJ  = sJ'
+  | otherwise                           = sJ
 
+filterBySearchBox :: [JSS.JSString] -> JSS.JSString -> [JSS.JSString]
+filterBySearchBox pkgs sJ
+  | JSS.isInfixOf "^" sJ = filter (JSS.isPrefixOf (stripSearch sJ)) pkgs
+  | JSS.isInfixOf "$" sJ = filter (JSS.isSuffixOf (stripSearch sJ)) pkgs
+  | otherwise            = filter (JSS.isInfixOf (stripSearch sJ))pkgs
+
+splitInfixPkg :: JSS.JSString -> JSS.JSString -> (T.Text, T.Text, T.Text)
+splitInfixPkg stripSJ pkg = (frontT, midT, backT)
+  where
+    textS               = JSS.textFromJSString stripSJ
+    (frontT, reminderT) = T.breakOn textS (JSS.textFromJSString pkg)
+    (midT, backT)       = T.breakOnEnd textS reminderT
+
+calcMatchesNew :: [JSS.JSString] -> JSS.JSString -> Matches
+calcMatchesNew pkgs sJss = 
+  if JSS.length sJss < 3 
+  then matchesEmpty
+  else Matches { matchesInput = textS, matchesExact = exactMap, matchesInfix = othersMap}
+  where
+    textS                = JSS.textFromJSString sJss
+    stripSJ              = stripSearch sJss
+    (exactPkg, infixPkg) = List.partition (== stripSJ) pkgs
+    exactMap             = (Map.fromList . fmap (\p -> (JSS.textFromJSString p,()))) exactPkg
+    othersMap            = (Map.fromList . fmap (\p -> (JSS.textFromJSString p, splitInfixPkg stripSJ p))) (filterBySearchBox infixPkg sJss)
```

Comment:
This feels like it is probably slower than before because you do two passovers (one with the partition and one with the othersMap) the whole package set. 

If you make another function call calcMatch :: Text -> Text -> (Map Text (),Map Text (Text,Text,Text))

you could make exactMap and othersMap with a single pass like:
```
(exactMap,othersMap) <- foldMap (calcMatch textS) pkgs
```

That may be nicer and faster.

Reply:
it is faster, indeed. I have put it in my new commit. Thanks for your input!!